docs/architecture.md

```diff
@@ -0,0 +1,271 @@
+
+Plugin functions get attached to the `Context` by the Rust technique
+of defining a trait and then implementing it for the `Context`. Thus,
+you get a global property value like so:
+
+```
+let b = context.get_global_property_value::<PropertyB>().unwrap();
+```
+
+The somewhat confusing Rust "turbofish" ::&lt;&gt; construction
+is an artifact of the way that global properties are stored. Specifically,
+each property is associated with a Rust type, which is also the
+lookup key for the property. So in this case, we are retrieving
+the value associated with the type `PropertyB` which will also
+be of type `PropertyB`. `get_global_property_value()` is a generic
+and the `::<PropertyB>` tells Rust to instantiate it for
+`PropertyB`.
+
+
+## Components
+
+A component is responsible for executing the logic of some piece of
+the simulation. Components are a very flexible mechanism, with the
+only requirement being that a component provides an `init()` function
+which is invoked when the component is first added to the `Context`.
+
+It is possible for a component to do all of its work inside `init()`.
+For instance, if you had a component which loaded a population into
+the simulation, that component would probably do that in `init()`
+because you want the population loaded prior to the simulation
+beginning. More complex components will want to do work *during* the
+simulation not just prior to it. In order to do so, they register one
+or more `Plan`s, which do the work of the simulation.
+
+
+## Scheduling Plans
+
+A `Plan` is a callback which is executed at some deterministic point
+of time in the future. For instance, if we wanted to model COVID
+isolation guidance, when someone is infected we might set a `Plan` to
+have them take a rapid antigen test 48 hours after becoming
+symptomatic.
+
+In order to schedule a `Plan` a component provides:
+
+* The *execution time* which is when the plan will execute
+* The *callback* to fire
+
+More details about callbacks are provided in [TODO] but at a high
+level, callbacks are Rust functions (typically closures) which
+get passed a reference to the `Context` as an argument. This
+allows them to act on the `Context` or any of its attached
+objects, such as data containers provided by plugins.
+
+A `Plan`'s callback can schedule other plans, thus allowing
+the simulation to condition future events on current events.
+In general, a `Component` will just work out the next action
+for a specific subject (e.g., a person in the simulation)
+and set a `Plan` for that. Once that action has taken place,
+the `Component` will then work out what happens next; it
+is generally not desirable to work out a detailed list
+of actions for each subject well in advance, as conditions
+may change rendering that planning obsolete.
+
+It is possible to schedule two `Plan`s for the same execution
+time. In this case, Eosim will run them in the order scheduled;
+in the future we might provide mechanisms to break this kind
```

Comment:
Consider saying 'provide mechanisms to control priority of events that are scheduled for the same time'
Or, we could drop this detail as it's minor and not currently implemented.